src/ceph-volume/ceph_volume/devices/lvm/prepare.py

```diff
@@ -169,6 +170,19 @@ def setup_device(self, device_type, device_name, tags):
             tags['ceph.%s_uuid' % device_type] = uuid
             tags['ceph.%s_device' % device_type] = path
             lv.set_tags(tags)
+        elif disk.is_device(device_name):
```

Comment:
this is disregarding the possibility that the disk might have LVs already in it, which might mean enough space is available (or not!). Creating the LVs is tremendously difficult and is why I've avoided it as much as I could. Since this feature is being pursued then all the smarts about sizing, and availability must be implemented.

Further, other systems like ceph-ansible might pass the same device over and over again, so some sort of idempotency has to exist.

Reply:
I'm not sure I follow. You mean a user passes a device containing an LV and there’s an expectation the LV is used? This is certainly not the intention and I hope nothing suggests things would work like that.

The sizing functionality will certainly be added here, but for the moment I think the functionality is valid. Idempotency is the ultimate goal of this refactoring, especially for the batch subcommand.

Reply:
There are several scenarios that could happen when a user passes a raw device like /dev/sda:

* device contains one or more LVs, user wants to use everything that is free 
* device contains one or more LVs, user wants to use a subset of what is free
* device contains one or more LVs and regardless of choice (if any) there is no space left
* device does not contain any LVs but user wants to use a specific size or percentage that can fit
* device does not contain any LVs and user's size would still not fit (e.g. 1TB LV in a 500gb device)

Calculating, resolving conflicts, and reporting back are going to be tremendously difficult - but it seems from your comment that will be coming at a later point?

Reply:
>     * device contains one or more LVs, user wants to use everything that is free

This is the current behaviour. A VG is created (if there is space) or re-used (if vg_name starts with ceph) and 100%FREE is used for the LV.

>     * device contains one or more LVs, user wants to use a subset of what is free
> 
>     * device contains one or more LVs and regardless of choice (if any) there is no space left

This will fail in any case.

>     * device does not contain any LVs but user wants to use a specific size or percentage that can fit
> 
>     * device does not contain any LVs and user's size would still not fit (e.g. 1TB LV in a 500gb device)

These cases will be covered in a future PR. I think the current behaviour is consistent with what prepare offers at the moment so I don't see any harm in it. After the sizing arguments are added, batch will use this code path which should help robustness and future development.

Reply:
Could you point me where is the validation to check for "if there is space" ? I don't see it. I couldn't find any checks for making sure that the user ask is going to be possible, and in the case it isn't, what the error messages are.

Reply:
Currently `create_lv` will use `100%FREE`. Since the pe_size of all VGs is currently 1GB this effectively limits the resulting LV size, no validation necessary. And since c-v allows any partition to passed regardless of size, the behaviour is consistent.